File: frontend/src/api/ci_v2.ts
/**
 * CIE v2.0 — 前端 API 调用层
 *
 * 对应后端: backend/app/api/ci_v2.py (7个端点)
 * 全类型注解, 无 any
 */
import api from './index'

// ── 类型定义 ───────────────────────────────────────

export interface RiskScoreData {
  ecr_id: number
  risk_score: number
  risk_level: string
  risk_vector: Record<string, number>
  mitigation_suggestions: string[]
  created_at: string
}

export interface ImpactNode {
  id: string
  node_type: string
  label: string
  impact_score: number
  affected_objects: Record<string, unknown>[]
  depth: number
}

export interface ImpactEdge {
  source_id: string
  target_id: string
  weight: number
  label: string
}

export interface ImpactGraphData {
  nodes: ImpactNode[]
  edges: ImpactEdge[]
  ripple_score: number
  max_depth: number
  node_count: number
  edge_count: number
}

export interface RecommendationData {
  recommendation: string
  required_approvers: string[]
  reason: string
  confidence: number
  risk_level: string
  risk_score: number
}

export interface ModelWeightsItem {
  version_id: string
  weights: Record<string, number>
  sample_count: number
  is_active: boolean
  created_at: string
}

export interface ModelParamsData {
  active: ModelWeightsItem | null
  history: ModelWeightsItem[]
}

// ── API 函数 ───────────────────────────────────────

/**
 * 查询 ECR 的风险评分
 * GET /api/v2/risk/{ecr_id}
 */
export function fetchRiskScore(ecrId: number): Promise<RiskScoreData> {
  return api.get(`/api/v2/risk/${ecrId}`).then(res => res.data as RiskScoreData)
}

/**
 * 查询 ECR 的变更影响图
 * GET /api/v2/impact-graph/{ecr_id}
 */
export function fetchImpactGraph(ecrId: number): Promise<ImpactGraphData> {
  return api.get(`/api/v2/impact-graph/${ecrId}`).then(res => res.data as ImpactGraphData)
}

/**
 * 查询 ECR 的审批推荐
 * GET /api/v2/approval-recommendation/{ecr_id}
 */
export function fetchApprovalRecommendation(ecrId: number): Promise<RecommendationData> {
  return api.get(`/api/v2/approval-recommendation/${ecrId}`)
    .then(res => res.data as RecommendationData)
}

/**
 * 批量风险评分 (最多20个)
 * POST /api/v2/risk/batch
 */
export function fetchBatchRisk(ecrIds: number[]): Promise<(RiskScoreData | null)[]> {
  return api.post('/api/v2/risk/batch', { ecr_ids: ecrIds })
    .then(res => res.data as (RiskScoreData | null)[])
}

/**
 * 提交预测结果反馈
 * POST /api/v2/feedback
 */
export function submitFeedback(ecrId: number, outcome: string, detail?: Record<string, unknown>): Promise<Record<string, unknown>> {
  return api.post('/api/v2/feedback', { ecr_id: ecrId, actual_outcome: outcome, outcome_detail: detail || null })
    .then(res => res.data)
}

/**
 * 查询模型参数
 * GET /api/v2/model-params
 */
export function fetchModelParams(): Promise<ModelParamsData> {
  return api.get('/api/v2/model-params').then(res => res.data as ModelParamsData)
}

/**
 * 回滚模型参数到指定版本
 * POST /api/v2/model-params/{version_id}
 */
export function rollbackModelParams(versionId: string): Promise<ModelWeightsItem> {
  return api.post(`/api/v2/model-params/${versionId}`)
    .then(res => res.data as ModelWeightsItem)
}

// ── Digital Thread — 事件链 ──────────────────────────

export interface EventChainItem {
  id: number
  event_type: string
  aggregate_type: string
  aggregate_id: number
  correlation_id: string
  causation_id: number | null
  event_data: string | null
  producer: string
  created_at: string
}

/**
 * 获取某个聚合 (ECR/ECO) 的完整事件链
 * GET /api/v2/event-graph/{aggregate_type}/{aggregate_id}
 */
export function fetchEventChain(aggregateType: 'ecr' | 'eco', aggregateId: number): Promise<EventChainItem[]> {
  return api.get(`/api/v2/event-graph/${aggregateType}/${aggregateId}`)
    .then(res => res.data as EventChainItem[])
}

/**
 * 获取因果链回溯（从最新事件沿 causation_id 回溯）
 * GET /api/v2/event-graph/{aggregate_type}/{aggregate_id}/causation
 */
export function fetchCausationChain(aggregateType: 'ecr' | 'eco', aggregateId: number): Promise<EventChainItem[]> {
  return api.get(`/api/v2/event-graph/${aggregateType}/${aggregateId}/causation`)
    .then(res => res.data as EventChainItem[])
}

File: app/schemas/ci_v2.py
"""CIE v2.0 — Pydantic Schema 定义

包含 M1 Risk Engine + M2 Impact Graph + M3 Approval Advisor 的全部 Schema。
全类型注解，无 Any。
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


# ════════════════════════════════════════════════════
# M1 — Risk Engine Schema
# ════════════════════════════════════════════════════

class RiskLevelEnum(str, Enum):
    """风险等级"""
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class RecommendationEnum(str, Enum):
    """推荐处理方式"""
    AUTO_APPROVE = "AUTO_APPROVE"
    FAST_TRACK = "FAST_TRACK"
    FULL_APPROVAL = "FULL_APPROVAL"
    REJECT_REDESIGN = "REJECT_REDESIGN"


class SignalInput(BaseModel):
    """风险信号输入 — 5个维度，范围0-100

    所有字段默认为0，支持输入缺失时的优雅降级。
    """
    bom_impact: float = Field(default=0, ge=0, le=100, description="BOM影响度")
    cert_impact: float = Field(default=0, ge=0, le=100, description="认证影响度")
    proto_instability: float = Field(default=0, ge=0, le=100, description="样机不稳定性")
    cost_overrun: float = Field(default=0, ge=0, le=100, description="成本超支风险")
    hist_failure_rate: float = Field(default=0, ge=0, le=100, description="历史故障率")

    model_config = ConfigDict(from_attributes=True)


class RiskAssessmentOut(BaseModel):
    """风险评估完整输出"""
    id: int
    ecr_id: int
    risk_score: float
    risk_level: RiskLevelEnum

    # 5个信号维度分
    bom_impact: float = 0
    cert_impact: float = 0
    proto_instability: float = 0
    cost_overrun: float = 0
    hist_failure_rate: float = 0

    signal_details: Optional[dict] = None
    mitigation_suggestions: Optional[list[str]] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ════════════════════════════════════════════════════
# M2 — Impact Graph Schema
# ════════════════════════════════════════════════════

class ImpactNode(BaseModel):
    """DAG 节点 — 代表变更传播链上的一个环节"""
    id: str
    node_type: str
    label: str
    impact_score: float
    affected_objects: list[dict]
    depth: int
    model_config = ConfigDict(from_attributes=True)


class ImpactEdge(BaseModel):
    """DAG 边 — 代表变更传播关系"""
    source_id: str
    target_id: str
    weight: float
    label: str
    model_config = ConfigDict(from_attributes=True)


class ImpactGraphOut(BaseModel):
    """变更影响图完整输出"""
    nodes: list[ImpactNode]
    edges: list[ImpactEdge]
    ripple_score: float
    max_depth: int
    node_count: int
    edge_count: int
    model_config = ConfigDict(from_attributes=True)


class RipplePath(BaseModel):
    """从根节点到叶子节点的一条完整传播路径"""
    path: list[str]
    total_score: float
    node_count: int
    model_config = ConfigDict(from_attributes=True)


# ════════════════════════════════════════════════════
# M3 — Approval Advisor Schema（预留）
# ════════════════════════════════════════════════════

class ApprovalAdvisorAction(str, Enum):
    """审批推荐动作"""
    AUTO_APPROVE = "AUTO_APPROVE"
    FAST_TRACK = "FAST_TRACK"
    FULL_APPROVAL = "FULL_APPROVAL"
    REJECT_REDESIGN = "REJECT_REDESIGN"


class ApprovalRecommendation(BaseModel):
    """审批推荐输出"""
    recommendation: ApprovalAdvisorAction
    required_approvers: list[str]
    reason: str
    confidence: float
    risk_level: str
    risk_score: float
    model_config = ConfigDict(from_attributes=True)


# ════════════════════════════════════════════════════
# M5 — API Schema（预留）
# ════════════════════════════════════════════════════

class RiskAssessmentApiResponse(BaseModel):
    """风险评分 API 响应"""
    ecr_id: int
    risk_score: float
    risk_level: str
    risk_vector: dict
    mitigation_suggestions: list[str]
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ════════════════════════════════════════════════════
# M4 — Feedback Loop Schema
# ════════════════════════════════════════════════════

ACTUAL_OUTCOME_CHOICES = ["approved", "rejected", "bom_success", "bom_failure", "cancelled"]


class PredictionOutcomeCreate(BaseModel):
    """反馈提交入参"""
    ecr_id: int
    actual_outcome: str  # 会在 validator 中检查
    outcome_detail: Optional[dict] = None

    @field_validator("actual_outcome")
    @classmethod
    def validate_outcome(cls, v: str) -> str:
        if v not in ACTUAL_OUTCOME_CHOICES:
            raise ValueError(f"actual_outcome 必须是 {ACTUAL_OUTCOME_CHOICES} 之一, 收到: {v}")
        return v


class PredictionOutcomeOut(BaseModel):
    """反馈记录输出"""
    id: int
    ecr_id: int
    risk_score: float
    risk_level: str
    predicted_action: Optional[str] = None
    actual_outcome: str
    outcome_detail: Optional[dict] = None
    recorded_at: datetime
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ModelWeightsOut(BaseModel):
    """权重参数版本输出"""
    version_id: str
    weights: dict
    sample_count: int
    is_active: bool
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)
